pkg/storaged/client.js

```diff
@@ -594,6 +597,50 @@ client.older_than = function older_than(version) {
     return utils.compare_versions(this.manager.Version, version) < 0;
 };
 
+/* Direct mounting and unmounting
+ *
+ * We don't use UDisks2 for most of our mounting and unmounting in
+ * order to get better control over which entry from fstab is
+ * selected.  Once UDisks2 allows that control, we can switch back to
+ * it.
+ *
+ * But note that these functions still require there to be an entry in
+ * fstab.
+ */
+
+client.mount_at = (block, target) => {
+    const entry = utils.array_find(block.Configuration,
+                                   c => c[0] == "fstab" && utils.decode_filename(c[1].dir.v) == target);
+    if (entry)
+        return cockpit.script('mkdir -p "$2" && mount "$1" "$2" -o "$3"',
```

Comment:
Why the mkdir? Does udisks do that? ISTM that the mount should just fail if the directory does not exist. Aside from being "too magic", there are also at least three traps here:

 - If the path contains a broken symlink, directory will be created (and possibly fails) in a possibly unexpected location.
 - `false && true` in POSIX shell is true (i.e. does not fail). You have to use `set -e; mkdir; mount` to get what you intend.
 - this will fail on funny characters in directory names (such as `"` or `\`). Shell is always awkward for this kind of thing.

In summary: Scrap the mkdir and the shell stuff, and use cockpit.spawn() for this.

Reply:
> Why the mkdir? Does udisks do that? 

Yes.

Reply:
> `false && true` in POSIX shell is true (i.e. does not fail).

This is not true (hah!) as far as I can tell.  If 'mkdir' fails, 'mount' will not be executed and cockpit.spawn will reject with an error. `bash -c "false && true"; echo $?` => `1`

> this will fail on funny characters in directory names (such as `"` or `\`)

I don't think so, since the `"$1"` construct is evaluated by bash, it is not done via cockpit.format or similar.  This works as expected and creates a directory named `f"o"o`  `bash -c 'mkdir "$1"' -- 'f"o"o'`

Reply:
Ah right: `false && true` will exit with 1, but only because it's the last exit code, not because the statement as a whole fails; `false && true; echo hello` will both run the echo *and* exit 0. I.e. these are a time bomb, as soon as someone attaches another command to that shell script it will explode. I maintain that `&&` is unintuitive and misleading, and `set -e` is more explicit and predictable.

TIL about the shell auto-magic quoting behaviour, so I give that to you. I still *hate* involving a shell in this, and I'd much rather have two chained cockpit.spawn() commands without a shell, but I won't veto it then (after moving to `set -e`, that one I insist on, sorry).

Reply:
Alright!

Reply:
Thanks! WFM now, resolving. Let's land the other PR first.